### PR TITLE
Feature public extensions

### PR DESCRIPTION
**Pull request by huboneo**
Screenshot 2020-10-26 at 15 35 56

### Please check if the PR fulfills these requirements
- [x] The commit message follows our [guidelines](https://github.com/neo4j-devtools/relate/blob/master/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?
This PR removes the use of a private JFrog NPM registry in favor of the public NPM registry. A relate extension is now defined as any NPM package with the `neo4j-relate-extension` keyword. Extensions published under the `@relate/` NPM namespace are marked as "official", and all signed extensions are verified against the [Code Signing CRL Server](https://github.com/neo4j-devtools/code-signing-crl-server).

This PR also adds the ability to install extensions from a URL.

### What is the current behavior?
Only extensions published on the private registry can be installed

### What is the new behavior?
Any extensions published on the public registry can be installed

### Does this PR introduce a breaking change?
Yes, old extensions can no longer be used

### Other information:
This PR uses Neo4j Desktop CA certificate for now, we need to issue a new intermediate certificate for relate (i'd assume). Further we should review how extension manifests are shaped and used, making sure we're not painting ourselves into a corner here

### CODE REVIEW

File: packages/common/src/entities/extensions/extensions.constants.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
+// @todo: Create cert for relate and decide where it should live
```
Comment (huboneo) [inline packages/common/src/entities/extensions/extensions.constants.ts:1]: This needs to be replaced with a new chain for relate

File: packages/common/src/entities/extensions/extensions.local.ts [hunk 2]
```diff
@@ -162,17 +171,15 @@ export class LocalExtensions extends ExtensionsAbstract<LocalEnvironment> {
...
         const execute = promisify(exec);
 
-        try {
+        if (extension.type !== EXTENSION_TYPES.STATIC) {
             await emitHookEvent(
                 HOOK_EVENTS.RELATE_EXTENSION_DEPENDENCIES_INSTALL_START,
                 `installing dependencies for ${extension.name}`,
             );
-            const output = await execute('npm install --production', {
+            const output = await execute('npm ci --production', {
```
Comment (huboneo) [inline packages/common/src/entities/extensions/extensions.local.ts:179]: we need to use CI to enable signed extensions to install dependencies, which means they have to be signed with their prod deps

File: packages/common/src/utils/extensions/extension-versions.ts [hunk 3]
```diff
@@ -61,71 +49,67 @@ export async function discoverExtension(extensionRootDir: string): Promise<IExte
     }
 
     const extensionManifest = path.join(extensionRootDir, EXTENSION_MANIFEST_FILE);
-    const extensionManifestLegacy = path.join(extensionRootDir, EXTENSION_MANIFEST_FILE_LEGACY);
-    const hasManifest = await fse.pathExists(extensionManifest);
-    const hasManifestLegacy = await fse.pathExists(extensionManifest);
+    const hasManifestFile = await fse.pathExists(extensionManifest);
     const extensionPackageJson = path.join(extensionRootDir, PACKAGE_JSON);
     const hasPackageJson = await fse.pathExists(extensionPackageJson);
```
Comment (huboneo) [inline packages/common/src/utils/extensions/extension-versions.ts:55]: Changes here are unrelated to the PR topic, just wanted to clean up how we detect extension meta

File: packages/common/src/utils/extensions/get-installed-extensions.ts [hunk 4]
```diff
@@ -58,15 +58,15 @@ function mapContentsToExtension(fullPath: string): IInstalledExtension {
...
+            name: _.replace(manifest.name, RELATE_NPM_PREFIX, ''),
             root: fullPath,
         });
     }
 
     if (hasPackageJson) {
         return new ExtensionModel({
             ...contents,
-            name: _.replace(contents.name, EXTENSION_NPM_PREFIX, ''),
+            name: _.replace(contents.name, RELATE_NPM_PREFIX, ''),
```
Comment (huboneo) [inline packages/common/src/utils/extensions/get-installed-extensions.ts:69]: not sure we should keep these replaces, I believe we did it for legibility, WDYT @nglgzz

Comment (nglgzz) [reply packages/common/src/utils/extensions/get-installed-extensions.ts:69]: It's probably best to remove it since it could cause ambiguity with packages published outside the `@relate` namespace.

**Comment by huboneo** [comment]
My one concern here is that we check the signature of an extension every time it is discovered (i.e. listed). This enables us to catch revocations on the fly, but slows down list operations due to network latency. WDYT @oskarhane ?

File: packages/common/src/entities/extensions/extensions.local.ts [hunk 5]
```diff
@@ -82,38 +81,46 @@ export class LocalExtensions extends ExtensionsAbstract<LocalEnvironment> {
...
 
-        const {extensionsCache} = this.environment.dirPaths;
+        const {extensionsCache, tmp} = this.environment.dirPaths;
+        const isFilePath = (await fse.pathExists(version)) && (await fse.stat(version)).isFile();
 
-        // version as a file path.
-        if ((await fse.pathExists(version)) && (await fse.stat(version)).isFile()) {
-            // extract extension to cache dir first
+        // version as a file path or URL.
+        if (isFilePath || isValidUrl(version)) {
```
Comment (jk05) [inline packages/common/src/entities/extensions/extensions.local.ts:88]: nice

File: packages/common/src/entities/extensions/extensions.local.ts [hunk 6]
```diff
@@ -82,38 +81,46 @@ export class LocalExtensions extends ExtensionsAbstract<LocalEnvironment> {
...
+                await emitHookEvent(HOOK_EVENTS.RELATE_EXTENSION_DOWNLOAD_STOP, null);
+            }
+
+            // extract extension to tmp first
+            const tmpExtractPath = path.join(tmp, uuidv4());
             const {name: extensionName, dist, version: extensionVersion} = await extractExtension(
-                version,
-                extensionsCache,
+                toExtract,
+                tmpExtractPath,
```
Comment (jk05) [inline packages/common/src/entities/extensions/extensions.local.ts:105]: Would we clean up this `tmpExtractPath` or did we say we would leave this to the OS to handle it for tmp?

File: packages/common/src/entities/extensions/extensions.local.ts [hunk 7]
```diff
@@ -82,38 +81,46 @@ export class LocalExtensions extends ExtensionsAbstract<LocalEnvironment> {
...
+            await fse.move(dist, cacheDir, {
                 overwrite: true,
             });
 
             try {
-                const discovered = await discoverExtension(destination);
+                const discovered = await discoverExtension(cacheDir);
 
-                return this.installRelateExtension(discovered, discovered.dist);
+                return this.installRelateExtension(discovered, discovered.dist).finally(() => fse.remove(cacheDir));
```
Comment (jk05) [inline packages/common/src/entities/extensions/extensions.local.ts:117]: guess this prevents building up cached extensions over time which is good

File: packages/common/src/entities/extensions/extensions.constants.ts (hunk 1 above)
Comment (huboneo) [reply packages/common/src/entities/extensions/extensions.constants.ts:1]: Alternatively not have it comitted at all, rn relate has no hardcoded secrets

File: packages/common/src/utils/extensions/extension-versions.ts [hunk 8]
```diff
@@ -135,62 +119,25 @@ export interface IExtensionVersion {
...
-                            name: `${extName}`,
-                            origin: EXTENSION_ORIGIN.ONLINE,
-                            version: `${version}`,
-                        };
-                    });
-                });
+        .mapEach(({name, origin, version}) => {
+            return {
+                name,
+                origin: origin || EXTENSION_ORIGIN.ONLINE,
```
Comment (nglgzz) [inline packages/common/src/utils/extensions/extension-versions.ts:138]: Wouldn't this be always `EXTENSION_ORIGIN.ONLINE`? And in what cases do we have a value coming from NPM that matches our enum?

File: packages/common/src/models/extension.model.ts [hunk 9]
```diff
@@ -27,3 +27,56 @@ export class ExtensionModel extends ModelAbstract<IInstalledExtension> implement
     @IsString()
     root!: string;
 }
+
+export class InstalledExtensionModel extends ModelAbstract<IInstalledExtension> implements IInstalledExtension {
+    @IsString()
+    name!: string;
```
[the remark this answers is not in the document]
Comment (huboneo) [reply packages/common/src/models/extension.model.ts:33]: Tells typescript to skip null check

**Comment by huboneo** [comment]
@nglgzz thank you for your feedback, I have updated the PR

File: packages/common/src/utils/extensions/extension-versions.ts (hunk 8 above)
Comment (huboneo) [reply packages/common/src/utils/extensions/extension-versions.ts:138]: No, this is mapping over the concatenated list of cached and online

File: packages/cli/docs/extension.md [hunk 10]
```diff
@@ -17,7 +17,7 @@ USAGE
   $ relate extension:install [NAME]
 
 OPTIONS
-  -V, --version=version          Version to install
+  -V, --version=version          Version to install (semver), or path to tarball
```
Comment (oskarhane) [inline packages/cli/docs/extension.md:20]: 👍

File: packages/cli/src/modules/extension/install.module.ts [hunk 11]
```diff
@@ -80,6 +83,11 @@ export class InstallModule implements OnApplicationBootstrap {
             const maybeWithName = name
                 ? _.filter([...choices.values()], (v) => v.name === name)
                 : [...choices.values()];
+
+            if (!arrayHasItems(maybeWithName)) {
+                throw new InvalidArgumentError(`Could not find installable extension with name: ${name}`);
+            }
```
Comment (oskarhane) [inline packages/cli/src/modules/extension/install.module.ts:89]: Great!

File: packages/common/src/constants.ts [hunk 12]
```diff
@@ -184,3 +183,11 @@ export enum FILTER_CONNECTORS {
...
 export const HEALTH_BASE_ENDPOINT = '/health';
 export const STATIC_APP_BASE_ENDPOINT = '/static';
+
+export enum EXTENSION_VERIFICATION_STATUS {
+    UNKNOWN = 'UNKNOWN',
+    UNSIGNED = 'UNSIGNED',
+    TRUSTED = 'TRUSTED',
+    UNTRUSTED = 'UNTRUSTED',
+    REVOKED = 'REVOKED',
+}
```
Comment (oskarhane) [inline packages/common/src/constants.ts:193]: Awesome that you added verification to packages 🎍

**Review by oskarhane — APPROVED** [review]
I ran into some issues that I pointed out on the side and you fixed, thanks!

First, this is great, this is an important step.

I do think we will run into issues with conflicts with app-ids quite fast though, so we should think about how to solve that.
The use case is being able to install multiple versions of the same ext and something we had a lot of feedback from in Desktop when we launched with the same limitations.

Maybe out of scope for this PR though.

File: packages/common/src/utils/extensions/extension-versions.ts (hunk 8 above)
Comment (nglgzz) [reply packages/common/src/utils/extensions/extension-versions.ts:138]: Ah right, I missed that part, I thought it was just the response coming from npm.

**Review by nglgzz — DISMISSED** [review]
Looks good to me. Looking forward to seeing Relate extensions on NPM :sparkles:

**Review by jk05 — DISMISSED** [review]

File: package.json [hunk 13]
```diff
@@ -16,7 +16,7 @@
         "lint": "eslint . --ext .ts",
         "test:teardown": "node scripts/tests/teardown.js",
         "test:setup": "node scripts/tests/setup.js",
-        "lerna:test": "lerna run test --concurrency 1",
+        "lerna:test": "lerna run test",
```
Comment (nglgzz) [inline package.json:19]: I'm afraid this will increase flakiness. Not because our tests can't run concurrently (they should) but because of the intensive file operations in our tests.

File: packages/common/src/entities/dbmss/dbmss.local.ts [hunk 14]
```diff
@@ -122,8 +122,13 @@ export class LocalDbmss extends DbmssAbstract<LocalEnvironment> {
 
         // version as a file path.
         if ((await fse.pathExists(version)) && (await fse.stat(version)).isFile()) {
-            const {extractedDistPath} = await extractNeo4j(version, this.environment.dirPaths.dbmssCache);
-            return this.installNeo4j(name, this.getDbmsRootPath(), extractedDistPath, credentials);
+            const tmpPath = path.join(this.environment.dirPaths.tmp, uuidv4());
+            const {extractedDistPath} = await extractNeo4j(version, tmpPath);
+            const dbms = await this.installNeo4j(name, this.getDbmsRootPath(), extractedDistPath, credentials);
+
+            await fse.remove(tmpPath);
```
Comment (nglgzz) [inline packages/common/src/entities/dbmss/dbmss.local.ts:129]: :+1:

File: package.json (hunk 13 above)
Comment (huboneo) [reply package.json:19]: I just ran 10 builds in a row without failures, so I think it's ok? we shouldnt be manipulating the same file at the same time?

**Review by nglgzz — APPROVED** [review]

**Comment by oskarhane** [comment]
> My one concern here is that we check the signature of an extension every time it is discovered (i.e. listed). This enables us to catch revocations on the fly, but slows down list operations due to network latency. WDYT @oskarhane ?

Good call. Let's start like this and optimize in the future if needed.
